pex/environment.py

```diff
@@ -27,19 +31,73 @@
 
 
 class PEXEnvironment(Environment):
+  class _CachingZipImporter(object):
+    class _CachingLoader(object):
+      def __init__(self, delegate):
+        self._delegate = delegate
+
+      def load_module(self, fullname):
+        loaded = sys.modules.get(fullname)
+        # Technically a PEP-302 loader should re-load the existing module object here - notably
+        # re-exec'ing the code found in the zip against the existing module __dict__. We don't do
+        # this since the zip is assumed immutable during our run and this is enough to work around
+        # the issue.
+        if not loaded:
+          loaded = self._delegate.load_module(fullname)
+          loaded.__loader__ = self
+        return loaded
+
+    _REGISTERED = False
+
+    @classmethod
+    def _ensure_namespace_handler_registered(cls):
+      if not cls._REGISTERED:
+        import pkg_resources
```

Comment:
Same question on these imports—is it required to be here to work, or here for readability? There are a couple of imports below as well like this.

Reply:
This was different and was delayed to only happen when we new namespace packages were in-play and thus setuptools available. It looks like this necessity linkage is now lost... looking at a refactor to re-establish.

Reply:
Re-established the link with better comments.

Reply:
Unfortunately, the link is really still just established by (better) comments. I could not re-establish a direct code link due to early-loading requirements of PEXEnvironments when many are involved (PEX_PATH). That case is particularly important since its needed for the original proximal cause of this bug this fixes in PEX.